gnovm/pkg/gnolang/transpile_gno0p9.go

```diff
@@ -0,0 +1,614 @@
+				// lastLine - 1 is the deleted crossing().
+				tf.MergeLine(lastLine - 1)
+				// and the next empty line too.
+				tf.MergeLine(lastLine)
+			}
+			didRemoveCrossing = false
+		}
+
+		// Main switch on c.Node() type.
+		switch gon := c.Node().(type) {
+		case *ast.Ident:
+			if gon.Name == "realm_XXX_TRANSPILE" {
+				// Impostor varname 'realm' will become
+				// renamed, so reclaim 'realm'.
+				gon.Name = "realm"
+			} else if gon.Name == "realm" {
+				// Rename name to _realm to avoid conflict with new builtin "realm".
+				// XXX: optimistic.
+				gon.Name = "_realm"
+			}
+		case *ast.ExprStmt:
+			if ce, ok := gon.X.(*ast.CallExpr); ok {
+				if id, ok := ce.Fun.(*ast.Ident); ok && id.Name == "crossing" {
+					// Validate syntax.
+					if len(ce.Args) != 0 {
+						err = errors.New("crossing called with non empty parameters")
+					}
+					// Delete statement 'crossing()'.
+					c.Delete()
+					didRemoveCrossing = true
+					setLast(gon.End())
+					return false
+				}
+			}
+		case *ast.FuncLit:
+			if inXforms2(gon, XTYPE_ADD_CURFUNC) {
+				gon.Type.Params.List = append([]*ast.Field{&ast.Field{
+					Names: []*ast.Ident{ast.NewIdent("cur")},
+					Type:  ast.NewIdent("realm_XXX_TRANSPILE"),
+				}}, gon.Type.Params.List...)
+				delete(xforms2, gon)
+			}
+		case *ast.FuncDecl:
+			if inXforms2(gon, XTYPE_ADD_CURFUNC) {
+				gon.Type.Params.List = append([]*ast.Field{&ast.Field{
+					Names: []*ast.Ident{ast.NewIdent("cur")},
+					Type:  ast.NewIdent("realm_XXX_TRANSPILE"),
+				}}, gon.Type.Params.List...)
+				delete(xforms2, gon)
+			}
+		case *ast.CallExpr:
+			if inXforms2(gon, XTYPE_ADD_CURCALL) {
+				gon.Args = append([]ast.Expr{ast.NewIdent("cur")}, gon.Args...)
+				delete(xforms2, gon)
+			} else if inXforms2(gon, XTYPE_ADD_NILREALM) {
+				gon.Args = append([]ast.Expr{ast.NewIdent("nil")}, gon.Args...)
+				delete(xforms2, gon)
+			}
+			if id, ok := gon.Fun.(*ast.Ident); ok && id.Name == "cross" {
+				// Replace expression 'cross(x)' by 'x'.
+				// In Gno 0.9 @cross decorator is used instead.
+				if len(gon.Args) == 1 {
+					c.Replace(gon.Args[0])
+				} else {
+					err = errors.New("cross called with invalid parameters")
+				}
+				return true
+			}
+		}
+		return true
+	}, nil)
+
+	// Verify that all xforms2 items were consumed.
+	if len(xforms2) > 0 {
+		for gon, xform2 := range xforms2 {
+			fmt.Println("left over:", xform2, gon)
+		}
+		panic("Xform items left over")
+	}
+
+	return err
+}
+
+// ========================================
+// WriteToMemPackage writes Go AST to a mempackage
+// This is useful for preparing prior version code for the preprocessor.
+func WriteToMemPackage(gofset *token.FileSet, gofs []*ast.File, mpkg *std.MemPackage) error {
+	for _, gof := range gofs {
+		fpath := gofset.File(gof.Pos()).Name()
+		_, fname := filepath.Split(fpath)
+		mfile := mpkg.GetFile(fname)
+		if mfile == nil {
+			if strings.HasPrefix(fname, ".") {
+				// Hidden files like .gnobuiltins.gno that
```

Comment:
I need to double-check, but are we correctly preventing people from having hidden files so that the only ones are the ones we control, such as .gnobuiltins.gno?

Btw, I think that we only have a .gnobuiltins.go, not anyone in .gno.

Reply:
There's nothing in the gofs gofset that can have a leading dot, so ".gnobuiltins.gno" is filtered out, and it filters out "." files in general as a convenience for future files like this.

gno.mod files don't have a leading dot. .gnobuiltins.gno is only for in memory type-checking.

Furthermore this is writing into an existing mpkg, overwriting Go files. Existing files with leading dots are not affected. I don't see the problem.